File: 21553 - Sai Nandan D/Week 3/106_Construct Binary Tree from Inorder and Postorder Traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def buildTree(self, inorder: List[int], postorder: List[int]) -> Optional[TreeNode]:
        inorder_map = {}
        for i, n in enumerate(inorder):
            inorder_map[n] = i
        # To be able to build the tree from the root, we've to start from the right end of `postorder`
        self.postorder_idx = -1
        
        def build(start, end):
            if start > end: return None            
            rootVal = postorder[self.postorder_idx]
            self.postorder_idx -= 1
            root = TreeNode(rootVal)
            # Build the right subtree first
            root.right = build(inorder_map[rootVal]+1, end)
            # And then the left subtree
            root.left = build(start, inorder_map[rootVal]-1)            
            return root
            
        return build(0, len(inorder)-1)

File: 21561 - Suvam Tamang/Week 3/Bonus/106_Construct Binary Tree from Inorder and Postorder Traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode buildTree(int[] inorder, int[] postorder) {
         if(postorder.length != inorder.length) 
             return null;
         return build(postorder, 0, postorder.length-1, inorder, 0, inorder.length-1);
    }
    
    public TreeNode build(int[] postorder, int postLow, int postHigh, int[] inorder, int inLow, int inHigh){
        if(postLow > postHigh || inLow > inHigh)
            return null;
        TreeNode root = new TreeNode(postorder[postHigh]);
        
        int inorderRoot = inLow;
        for(int i=inLow;i<=inHigh;i++){
            if(inorder[i]==root.val){ 
                inorderRoot=i; 
                break; 
            }
        }
        int leftTreeLen = inorderRoot-inLow;
        root.left = build(postorder, postLow, postLow+leftTreeLen-1, inorder, inLow, inorderRoot-1);
        root.right = build(postorder, postLow+leftTreeLen, postHigh-1, inorder, inorderRoot+1, inHigh);       
        return root;   
    }
}
